File: Common/SCFileWatcher.m
//
//  SCFileWatcher.m
//  SelfControl
//
//  Created by Charlie Stigler on 3/20/21.
//

#import "SCFileWatcher.h"
#include <CoreServices/CoreServices.h>

@implementation SCFileWatcher

static void SCFileWatcherGlobalCallback(
    ConstFSEventStreamRef streamRef,
    void *callbackCtxInfo,
    size_t numEvents,
    void *eventPaths, // CFArrayRef
    const FSEventStreamEventFlags eventFlags[],
    const FSEventStreamEventId eventIds[])
{
    NSArray* paths = (__bridge NSArray*)eventPaths;
    SCFileWatcher* watcher = (__bridge SCFileWatcher*)callbackCtxInfo;
    [watcher directoryWatcherTriggered: paths flags: eventFlags];
}

- (void)directoryWatcherTriggered:(NSArray<NSString*>*)eventPaths flags:(const FSEventStreamEventFlags[])eventFlags {
    BOOL triggerFileWatcher = NO;

    for (unsigned int i = 0; i < eventPaths.count; i++) {
        NSString* eventPath = [eventPaths[i] stringByStandardizingPath];
        
        if ([eventPath isEqualToString: self.filePath]) {
            triggerFileWatcher = YES;
        }
    }
    
    if (triggerFileWatcher) {
        self.callbackBlock(nil);
    }
}

+ (instancetype)watcherWithFile:(NSString*)watchPath block:(void(^)(NSError* error))callbackBlock {
    return [[SCFileWatcher new] initWithFile: watchPath block: callbackBlock];
}

- (instancetype)initWithFile:(NSString*)watchPath block:(void(^)(NSError* error))callbackBlock {
    self = [super init];

    NSFileManager* fileMan = [NSFileManager defaultManager];
    _filePath = [watchPath stringByStandardizingPath];
    BOOL isDirectory;
    [fileMan fileExistsAtPath: self.filePath isDirectory: &isDirectory];
    
    NSString* directoryPath;
    if (isDirectory) {
        directoryPath = self.filePath;
    } else {
        directoryPath = [self.filePath stringByDeletingLastPathComponent];
    }

    FSEventStreamContext callbackCtx;
    callbackCtx.version = 0;
    callbackCtx.info = (__bridge void *)self;
    callbackCtx.retain = NULL;
    callbackCtx.release = NULL;
    callbackCtx.copyDescription = NULL;

    FSEventStreamRef eventStream = FSEventStreamCreate(
        kCFAllocatorDefault,
        &SCFileWatcherGlobalCallback,
        &callbackCtx, // context
        (__bridge CFArrayRef)@[directoryPath],
        kFSEventStreamEventIdSinceNow,
        1.5, // seconds to throttle callbacks
        kFSEventStreamCreateFlagUseCFTypes | kFSEventStreamCreateFlagMarkSelf | kFSEventStreamCreateFlagIgnoreSelf | kFSEventStreamCreateFlagFileEvents
    );
    
    FSEventStreamScheduleWithRunLoop(eventStream,
                                     [[NSRunLoop currentRunLoop] getCFRunLoop],
                                     kCFRunLoopDefaultMode);
    if (!FSEventStreamStart(eventStream)) {
        NSLog(@"WARNING: failed to start watching file %@", watchPath);
        FSEventStreamUnscheduleFromRunLoop(eventStream, [[NSRunLoop currentRunLoop] getCFRunLoop], kCFRunLoopDefaultMode);
        FSEventStreamInvalidate(eventStream);
        FSEventStreamRelease(eventStream);
        return nil;
    }
    
    _eventStream = eventStream;
    _callbackBlock = callbackBlock;
    
    return self;
}

- (void)stopWatching {
    FSEventStreamStop(self.eventStream);
    FSEventStreamUnscheduleFromRunLoop(self.eventStream, [[NSRunLoop currentRunLoop] getCFRunLoop], kCFRunLoopDefaultMode);
    FSEventStreamInvalidate(self.eventStream);
    FSEventStreamRelease(self.eventStream);

    _eventStream = NULL;
}

@end

File: Common/Utilities/SCFileWatcher.swift
import Foundation
import CoreServices

/// Watches a file or directory for changes using FSEvents.
class SCFileWatcher {
    private var stream: FSEventStreamRef?
    private let path: String
    private let callback: () -> Void

    init(path: String, callback: @escaping () -> Void) {
        self.path = path
        self.callback = callback
    }

    func start() {
        let pathsToWatch = [path] as CFArray

        // [Fix #6] Use passRetained to prevent use-after-free if watcher is
        // deallocated before the stream is invalidated.
        var context = FSEventStreamContext()
        context.info = Unmanaged.passRetained(self).toOpaque()
        context.release = { info in
            guard let info = info else { return }
            Unmanaged<SCFileWatcher>.fromOpaque(info).release()
        }

        let flags: FSEventStreamCreateFlags = UInt32(
            kFSEventStreamCreateFlagFileEvents | kFSEventStreamCreateFlagUseCFTypes
        )

        guard let stream = FSEventStreamCreate(
            nil,
            { _, info, _, _, _, _ in
                guard let info = info else { return }
                let watcher = Unmanaged<SCFileWatcher>.fromOpaque(info).takeUnretainedValue()
                watcher.callback()
            },
            &context,
            pathsToWatch,
            FSEventStreamEventId(kFSEventStreamEventIdSinceNow),
            1.0, // latency in seconds
            flags
        ) else { return }

        self.stream = stream
        FSEventStreamScheduleWithRunLoop(stream, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
        FSEventStreamStart(stream)
    }

    func stop() {
        guard let stream = stream else { return }
        FSEventStreamStop(stream)
        FSEventStreamInvalidate(stream)
        FSEventStreamRelease(stream)
        self.stream = nil
    }

    deinit {
        stop()
    }
}
